File: employee-service-python/app/routes/employee_routes.py
from flask import Blueprint, request
from app.controllers.employee_controller import EmployeeController
from loguru import logger
import time

# Create a Blueprint for employee routes
employee_bp = Blueprint('employee', __name__)

# Middleware for logging request details
@employee_bp.before_request
def log_request_info():
    """Log request details before processing"""
    logger.info(f"Request: {request.method} {request.path}")
    # Store request start time for duration calculation
    request.start_time = time.time()

# Middleware for logging response details
@employee_bp.after_request
def log_response_info(response):
    """Log response details after processing"""
    duration = time.time() - request.start_time
    logger.info(f"Response: {response.status_code} - took {duration:.4f}s")
    return response

# Route: Get all employees
@employee_bp.route('', methods=['GET'])
def get_all_employees():
    """Get all employees endpoint"""
    return EmployeeController.get_all_employees()

# Route: Get employee by ID
@employee_bp.route('/<employee_id>', methods=['GET'])
def get_employee_by_id(employee_id):
    """Get employee by ID endpoint"""
    return EmployeeController.get_employee_by_id(employee_id)

# Route: Create a new employee
@employee_bp.route('', methods=['POST'])
def create_employee():
    """Create a new employee endpoint"""
    return EmployeeController.create_employee()

# Route: Update an employee
@employee_bp.route('/<employee_id>', methods=['PUT'])
def update_employee(employee_id):
    """Update an employee endpoint"""
    return EmployeeController.update_employee(employee_id)

# Route: Delete an employee
@employee_bp.route('/<employee_id>', methods=['DELETE'])
def delete_employee(employee_id):
    """Delete an employee endpoint"""
    return EmployeeController.delete_employee(employee_id)

File: employee-service-rust/src/routes/employee_routes.rs
use actix_web::web;
use std::time::Instant;
use tracing::info;

use crate::controllers::employee_controller;

// Configure employee routes
pub fn config(cfg: &mut web::ServiceConfig) {
    cfg.service(
        web::scope("/api/employees")
            // Add middleware for request logging
            .wrap(RequestTimer)
            // Register employee routes
            .service(employee_controller::get_all_employees)
            .service(employee_controller::get_employee_by_id)
            .service(employee_controller::create_employee)
            .service(employee_controller::update_employee)
            .service(employee_controller::delete_employee),
    );
}

// Request timer middleware
use actix_web::{
    dev::{self, Service, ServiceRequest, ServiceResponse, Transform},
    Error,
};
use futures::future::{ok, Ready};
use futures::Future;
use std::pin::Pin;
use std::task::{Context, Poll};

// Middleware for logging request timing
pub struct RequestTimer;

impl<S, B> Transform<S, ServiceRequest> for RequestTimer
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
    S::Future: 'static,
    B: 'static,
{
    type Response = ServiceResponse<B>;
    type Error = Error;
    type InitError = ();
    type Transform = RequestTimerMiddleware<S>;
    type Future = Ready<Result<Self::Transform, Self::InitError>>;

    fn new_transform(&self, service: S) -> Self::Future {
        ok(RequestTimerMiddleware { service })
    }
}

pub struct RequestTimerMiddleware<S> {
    service: S,
}

impl<S, B> Service<ServiceRequest> for RequestTimerMiddleware<S>
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
    S::Future: 'static,
    B: 'static,
{
    type Response = ServiceResponse<B>;
    type Error = Error;
    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>>>>;

    fn poll_ready(&self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.service.poll_ready(cx)
    }

    fn call(&self, req: ServiceRequest) -> Self::Future {
        // Log request details
        info!("Request: {} {}", req.method(), req.path());
        let start = Instant::now();

        let fut = self.service.call(req);

        Box::pin(async move {
            // Process request
            let res = fut.await?;
            
            // Log response details
            let duration = start.elapsed();
            info!(
                "Response: {} - took {:.4}s",
                res.status().as_u16(),
                duration.as_secs_f32()
            );
            
            Ok(res)
        })
    }
}
